src/index.ts

```diff
@@ -1568,44 +1565,61 @@ export class BigQuery extends common.Service {
    */
   query(
     query: string | Query,
-    optionsOrCallback?:
-      | QueryOptions
-      | SimpleQueryRowsCallback
-      | QueryRowsCallback,
-    cb?: SimpleQueryRowsCallback | QueryRowsCallback
-  ): void | Promise<SimpleQueryRowsResponse> | Promise<QueryRowsResponse> {
-    let options =
+    optionsOrCallback?: QueryOptions | QueryRowsCallback,
+    cb?: QueryRowsCallback
+  ): void | Promise<QueryRowsResponse> {
+    const options =
       typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
     const callback =
       typeof optionsOrCallback === 'function' ? optionsOrCallback : cb;
-    this.createQueryJob(query, (err, job, resp) => {
-      if (err) {
-        (callback as SimpleQueryRowsCallback)(err, null, resp);
-        return;
-      }
-      if (typeof query === 'object' && query.dryRun) {
-        (callback as SimpleQueryRowsCallback)(null, [], resp);
-        return;
-      }
-      // The Job is important for the `queryAsStream_` method, so a new query
-      // isn't created each time results are polled for.
-      options = extend({job}, options);
-      job!.getQueryResults(options, callback as QueryRowsCallback);
-    });
+
+    query = typeof query === 'string' ? {query} : query;
+
+    if (query.table) {
+      query.table.getRows(options, callback!);
+      return;
+    }
+
+    this.runQuery_(query, options).then(
+      ([rows, nextQuery, resp]) => callback!(null, rows, nextQuery, resp),
+      callback!
+    );
   }
 
   /**
-   * This method will be called by `createQueryStream()`. It is required to
-   * properly set the `autoPaginate` option value.
-   *
    * @private
    */
-  queryAsStream_(query: Query, callback?: SimpleQueryRowsCallback) {
-    if (query.job) {
-      query.job.getQueryResults(query, callback as QueryRowsCallback);
-      return;
+  async runQuery_(query: Query, options: QueryOptions): Promise<RowsResponse> {
+    const [job, resp] = await this.createQueryJob(query);
+
+    if (query.dryRun) {
+      return [[], null, resp];
+    }
+
+    const queryOptions = Object.assign({}, options, {autoPaginate: false});
+    let rows, nextQuery, results;
+
+    do {
+      [rows, nextQuery, results] = await job.getQueryResults(queryOptions);
+    } while (!results!.jobComplete);
+
+    if (!rows.length) {
+      return [rows, null, results!];
     }
-    this.query(query, {autoPaginate: false}, callback);
+
+    const {datasetId, tableId} = resp.configuration!.query!.destinationTable!;
+    const table = this.dataset(datasetId!).table(tableId!);
+
+    table.metadata = {schema: results!.schema};
+
+    const tableData = await table.getRows(options);
```

Comment:
All our promises return tuples, so `tableData` is essentially `[rows, nextPageToken, fullResponseBody]`

Reply:
It looks like for query dryrun this returns the job insert response as the third element, but for real queries the third element is the getQueryResults response?  This seems inconsistent, but I may be misunderstanding.

Reply:
No, that is exactly what you're seeing. This is a pattern you'll see in pretty much every Node client library, we always return the full response body as the last element. I agree that it can be a bit odd in methods that make several API requests.

@bcoe any feelings here?

Reply:
@callmehiphop @shollyman @bcoe  Any progress on this? It's an issue that I'm getting pinged on regarding the status.

Reply:
no super strong opinion, as long as we're consistent; If we had a do over, I'd probably return an object with meaningful keys names, rather than a tuple, which one could then destructure -- but I don't want to break our design pattern.